algorithms/scripts/RandomForestClassifier-mans2singh.py

```diff
@@ -0,0 +1,299 @@
+
+
+# In[220]:
+
+clf_grid.best_params_
+
+
+# In[221]:
+
+best_clf
+
+
+# ## Visualize hyperparameters performance
+
+# In[222]:
+
+def grid_scores_to_df(grid_scores):
+    """
+    Convert a sklearn.grid_search.GridSearchCV.grid_scores_ attribute to 
+    a tidy pandas DataFrame where each row is a hyperparameter-fold combinatination.
+    """
+    rows = list()
+    for grid_score in grid_scores:
+        for fold, score in enumerate(grid_score.cv_validation_scores):
+            row = grid_score.parameters.copy()
+            row['fold'] = fold
+            row['score'] = score
+            rows.append(row)
+    df = pd.DataFrame(rows)
+    return df
+
+
+# ## Process Mutation Matrix
+
+# In[223]:
+
+cv_score_df = grid_scores_to_df(clf_grid.grid_scores_)
+cv_score_df.head(2)
+
+
+# In[224]:
+
+# Cross-validated performance distribution
+facet_grid = sns.factorplot(x='max_depth', y='score', col='n_estimators',
+    data=cv_score_df, kind='violin', size=4, aspect=1)
+facet_grid.set_ylabels('AUROC');
+
+
+# In[225]:
+
+# Cross-validated performance heatmap
+cv_score_mat = pd.pivot_table(cv_score_df, values='score', index='max_depth', columns='n_estimators')
+ax = sns.heatmap(cv_score_mat, annot=True, fmt='.1%')
+ax.set_xlabel('(n_estimators)')
+ax.set_ylabel('(max_depth)');
+
+
+# ## Use Optimal Hyperparameters to Output ROC Curve
+
+# In[226]:
+
+y_pred_train = pipeline.predict_proba(X_train)[:, 1]
+y_pred_test = pipeline.predict_proba(X_test)[:, 1]
+
+def get_threshold_metrics(y_true, y_pred):
+    roc_columns = ['fpr', 'tpr', 'threshold']
+    roc_items = zip(roc_columns, roc_curve(y_true, y_pred))
+    roc_df = pd.DataFrame.from_items(roc_items)
+    auroc = roc_auc_score(y_true, y_pred)
+    return {'auroc': auroc, 'roc_df': roc_df}
+
+metrics_train = get_threshold_metrics(y_train, y_pred_train)
+metrics_test = get_threshold_metrics(y_test, y_pred_test)
+
+
+# In[227]:
+
+# Plot ROC
+plt.figure()
+for label, metrics in ('Training', metrics_train), ('Testing', metrics_test):
+    roc_df = metrics['roc_df']
+    plt.plot(roc_df.fpr, roc_df.tpr,
+        label='{} (AUROC = {:.1%})'.format(label, metrics['auroc']))
+plt.xlim([0.0, 1.0])
+plt.ylim([0.0, 1.05])
+plt.xlabel('False Positive Rate')
+plt.ylabel('True Positive Rate')
+plt.title('Predicting TP53 mutation from gene expression (ROC curves)')
+plt.legend(loc='lower right');
+
+
+# ## What are the classifier coefficients?
+
+# In[233]:
+
+best_clf.feature_importances_
+
+
+# In[236]:
+
+feature_importances_df = pd.DataFrame(best_clf.feature_importances_.transpose(), index=X.columns[feature_mask], columns=['weight'])
+feature_importances_df['abs'] = feature_importances_df['weight'].abs()
+feature_importances_df = feature_importances_df.sort_values('abs', ascending=False)
+
+
+# In[237]:
+
+'{:.1%} zero coefficients; {:,} negative and {:,} positive coefficients'.format(
```

Comment:
Since feature_importance is always positive, this cell can be removed.

Instead it may be nice to show the `feature_importance` distribution. Check out `seaborn.distplot`.